[Security Solution][Detections] Adds Bulk edit API

vitaliidm opened the pull request:
- Issue: https://github.com/elastic/kibana/issues/119600
## Summary

- Adds bulk update API on server side by introducing new bulk action `edit` to `/rules/_bulk_action` API. Draft documentation is available by https://github.com/elastic/security-docs/pull/1317
- Added unit and functional tests
- Added actions for editing: tags, index, timeline(the rest of required actions can be added when needed)
- Little cleanup of updateRules/patchRules typing(removed unused props)
- Limited concurrency to 5, for `/rules/_bulk_action` API
- Moved `security_solution/server/endpoint/routes/limited_concurrency.ts` to `security_solution/server/routes/limited_concurrency.ts `. Cleaned up unused constants in endpoint directory
- Refactored `registerLimitedConcurrencyRoutes` method:  rate limit is applying per route path now, rather than for all routes, as previously 
- Added error handling for single rule update in _bulk_action API, so if one rule action fails, it won't fail the whole request. Caught errors for single are returning in response
- Added chunking of concurrent update requests. Rules are split into chunks of 50 rules, rules within each chunk are processed in parallel, chunks are processed sequentially.

### Example
```console
curl --location --request POST 'http://localhost:5601/kbn/api/detection_engine/rules/_bulk_action' \
--header 'kbn-xsrf: reporting' \
--header 'Authorization: Basic PUT_ENCODED_USER:PASSWORD' \
--header 'Content-Type: application/json' \
--data-raw '{
    "action": "edit",
    "query": "",
    "edit":  [{
        "type": "add_tags",
        "value": ["sample_tag"]
    }
    ]
}'
```
where PUT_ENCODED_USER:PASSWORD equals to base64 encoded string 'user:password'

### Checklist

Delete any items that are not applicable to this PR.

- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

vitaliidm commented:
@elasticmachine merge upstream

vitaliidm commented:
@elasticmachine merge upstream

vitaliidm commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

vitaliidm commented:
@elasticmachine merge upstream

vitaliidm commented:
@elasticmachine merge upstream

vitaliidm commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

xcrzx commented on x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts:
```diff
@@ -447,12 +447,94 @@ export const privilege = t.type({
   'duplicate' = 'duplicate',
 }
 
+export const bulkActionSimple = enumeration('BulkActionSimple', BulkActionSimple);
+
+export enum BulkActionEnhanced {
+  'update' = 'update',
+}
+export const bulkActionEnhanced = enumeration('BulkActionEnhanced', BulkActionEnhanced);
+
+export const BulkAction = {
+  ...BulkActionSimple,
+  ...BulkActionEnhanced,
+};
+
+export type BulkAction = BulkActionSimple | BulkActionEnhanced;
```
Can we avoid splitting bulk action into multiple enums: `BulkActionSimple` and `BulkActionEnhanced`? I think the current approach is not easily extensible. For example, if we would need to add more actions, we had to add one more enum, as `BulkActionEnhanced` is designed to hold exactly one action: `update`, and adding more would break this validation:

```ts
t.type({
  query: queryOrUndefined,
  action: bulkActionEnhanced,
  updates: t.array(bulkActionUpdate),
})
```

vitaliidm replied:
I did not find a way to do this and keep validation correct.

I split it into 2 enums as we essentially have 2 types of actions:
1. simpler actions, with just 2 properties: action and query
2. update(enhanced) with 3 properties: action, query, updates.

So, in this case: `updates` property will be required for `update` action only, and will not be validated for other actions.

Initially, I did not split actions into 2 enums, but put updates property as partial. But downside in that approach was I had to do additional validation when processing actions in route function. And essentially that solution would not take full advantage of request's body validation we have.

> For example, if we would need to add more actions, we had to add one more enum, as BulkActionEnhanced is designed to hold exactly one action: update, and adding more would break this validation:

I think it would depend on type of action we want to introduce: if it does not require any additional properties it can be put into `BulkActionSimple`. If it would require another unique property, specific to this action, we would need to create a separate schema for this action to validate it.

xcrzx replied:
Just tried to do that with simple unions, validation is working fine:

```diff
diff --git a/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts b/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts
index 47e21487bf7..3ba26db1468 100644
--- a/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts
+++ b/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts
@@ -447,27 +447,14 @@ export const privilege = t.type({
 
 export type Privilege = t.TypeOf<typeof privilege>;
 
-export enum BulkActionSimple {
+export enum BulkAction {
   'enable' = 'enable',
   'disable' = 'disable',
   'export' = 'export',
   'delete' = 'delete',
   'duplicate' = 'duplicate',
-}
-
-export const bulkActionSimple = enumeration('BulkActionSimple', BulkActionSimple);
-
-export enum BulkActionEnhanced {
   'update' = 'update',
 }
-export const bulkActionEnhanced = enumeration('BulkActionEnhanced', BulkActionEnhanced);
-
-export const BulkAction = {
-  ...BulkActionSimple,
-  ...BulkActionEnhanced,
-};
-
-export type BulkAction = BulkActionSimple | BulkActionEnhanced;
 
 export const bulkAction = enumeration('BulkAction', BulkAction);
 
diff --git a/x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.ts b/x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.ts
index 4855087de4c..e4fcabd7868 100644
--- a/x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.ts
+++ b/x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.ts
@@ -6,24 +6,25 @@
  */
 
 import * as t from 'io-ts';
-import {
-  bulkActionEnhanced,
-  bulkActionSimple,
-  queryOrUndefined,
-  bulkActionUpdate,
-} from '../common/schemas';
+import { BulkAction, queryOrUndefined, bulkActionUpdate } from '../common/schemas';
 
 export const performBulkActionSchema = t.union([
   t.exact(
     t.type({
       query: queryOrUndefined,
-      action: bulkActionSimple,
+      action: t.union([
+        t.literal(BulkAction.delete),
+        t.literal(BulkAction.disable),
+        t.literal(BulkAction.duplicate),
+        t.literal(BulkAction.enable),
+        t.literal(BulkAction.export),
+      ]),
     })
   ),
   t.exact(
     t.type({
       query: queryOrUndefined,
-      action: bulkActionEnhanced,
+      action: t.literal(BulkAction.update),
       updates: t.array(bulkActionUpdate),
     })
   ),
```

vitaliidm replied:
updated, thank for suggestion

xcrzx commented on x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.test.ts:
```diff
@@ -41,12 +43,11 @@ describe('perform_bulk_action_schema', () => {
-    const message = foldLeftRight(checked);
     expect(getPaths(left(message.errors))).toEqual([
       'Invalid value "undefined" supplied to "action"',
+      'Invalid value "undefined" supplied to "updates"',
```
Why did this message change? I think the previous variant was more accurate for this test case.

xcrzx commented on x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.test.ts:
```diff
@@ -8,18 +8,22 @@
-import { BulkAction } from '../common/schemas';
+import { BulkAction, BulkActionUpdateType } from '../common/schemas';
+
+const retrieveValidationMessage = (payload: PerformBulkActionSchema) => {
```
I think `unknown` would be a more accurate type than `PerformBulkActionSchema` as we're testing random payload as input. So such castings as `payload as unknown as PerformBulkActionSchema` could be omitted.

vitaliidm replied:
This makes sense, thanks for the suggestion

xcrzx commented on x-pack/plugins/security_solution/server/lib/detection_engine/rules/bulk_action_update.ts:
```diff
@@ -0,0 +1,61 @@
+  Array.from(new Set([...arr, ...items]));
+
+export const deleteItemsFromArray = <T>(arr: T[], items: T[]): T[] => {
+  const itemsSet = new Set(items);
+  return arr.filter((item) => !itemsSet.has(item));
```
Why do you need `new Set(items)` here? Wouldn't `items.includes(item)` work?

vitaliidm replied:
I did this to avoid N*M time complexity, where N - existing tags, M - tags to remove.
As with `.include` it would've do search through the whole array of items on every filter callback call.

xcrzx replied:
I don't think users will be removing thousands of rules for the effect of this optimization to be visible 🙂  But ok, I'll leave it up to you

xcrzx commented on x-pack/test/detection_engine_api_integration/security_and_spaces/tests/perform_bulk_action.ts:
```diff
@@ -156,5 +145,51 @@ export default ({ getService }: FtrProviderContext): void => {
       expect(rulesResponse.total).to.eql(2);
     });
+
+    it('should set, add and delete tags in rules', async () => {
```
Should we add tests for index and timeline actions as well?

vitaliidm replied:
I have it covered in unit tests: https://github.com/elastic/kibana/blob/61efff8ee88e2a23d98a6618be0e24e6a4d070a5/x-pack/plugins/security_solution/server/lib/detection_engine/rules/bulk_action_update.test.ts#L74

Thought, it might be redundant to do it for functional tests, as I already covered update action(for tags) to confirm e2e running is successful. 
Do you think we need to a functional test for every action?

xcrzx replied:
Unit tests for the `appplyBulkActionUpdateToRule` method are good, but they don't cover other methods used by the bulk action API. For example, if something changes inside the `patchRules` method, how would it affect our bulk API? Unit tests cannot answer that question, as they test methods in isolation. On the other hand, functional tests test API as a whole and verify that we should expect to receive particular output for the given input no matter what changes inside the route. And that's what we need in this case.

vitaliidm replied:
added additional test

xcrzx commented on x-pack/test/detection_engine_api_integration/security_and_spaces/tests/perform_bulk_action.ts:
```diff
@@ -156,5 +145,51 @@ export default ({ getService }: FtrProviderContext): void => {
+          query: '',
+          action: BulkAction.update,
+          updates: [
+            {
+              type: BulkActionUpdateType.set_tags,
+              value: [],
+            },
+            {
+              type: BulkActionUpdateType.add_tags,
+              value: tags,
+            },
```
Can we test `set` and `add` updates separately? It is not completely clear what a test with two updates combined verifies.

xcrzx commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts:
```diff
@@ -20,10 +21,66 @@ import { deleteRules } from '../../rules/delete_rules';
+    if (rulesChunks.length === 0) {
+      return acc;
+    }
+    const processed = await Promise.all(
+      rulesChunks[0].map(async (rule) => actionPerformWrapper(async () => action(rule), rule))
+    );
```
I think using `Promise.all` is suboptimal when chunking outgoing requests. For example, let's say we're sending 50 requests in batches by 10. If one of the first ten requests takes too long to complete, it will hold back all consequent batches from being sent. So total time is 5 x (slowest request in each batch).

It could be optimized by sending a new request once one of the first ones has finished. See this PR for alternative implementation: https://github.com/elastic/kibana/pull/121110/files#diff-e34d09428e2b32ee8f5000236542eb0eee8674e6c8c1c7af9d91c2db4c5b181f

xcrzx commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts:
```diff
@@ -145,9 +203,43 @@ export const performBulkActionRoute = (
               body: responseBody,
             });
+          case BulkAction.update:
+            processed = await chunkifyRulesAction(rules.data, async (rule) => {
```
Should we also chunkify other bulk actions?

vitaliidm replied:
yes, but I've decided to implement it in one of the subsequent PRs, where I will add on UI handling errors object in response, when any of rule updates fail. 
That way it won't break anything in UX. 

Another reasoning:
In current implementation of other bulk actions, any failure would end request, so chunking the requests could postpone showing the error to user.

xcrzx replied:
> I will add on UI handling errors object in response, when any of rule updates fail. That way it won't break anything in UX.

Not sure I understand. How sending requests in chunks could break UX?

> In current implementation of other bulk actions, any failure would end request, so chunking the requests could postpone showing the error to user

That should be fine. We can show an error when all requests have finished.

vitaliidm replied:
> Not sure I understand. How sending requests in chunks could break UX?

Not sending in chunks, but returning 200 and errors array in response instead of failing the whole request. In my implementation chunking and error handling are coupled together in chunkify method.

> That should be fine. We can show an error when all requests have finished.

do you suggest to chunk requests for all actions, catch errors, and then if action is not `update` return error response?

xcrzx replied:
We should try making responses from different actions homogeneous. If one action returns 500 when any rule update fails, and the other returns 200 + an array of errors, we'll need additional logic on the front end to handle both cases. And to me, it looks like something we should avoid.

But do we need to respond with 200 if some of the rule updates fail?

vitaliidm replied:
> We should try making responses from different actions homogeneous. If one action returns 500 when any rule update fails, and the other returns 200 + an array of errors, we'll need additional logic on the front end to handle both cases. And to me, it looks like something we should avoid.

Yes, I agree, response should be similar. But we don't have yet FE for update, so no additional logic is required for now 
 and won't be required in future as work on bulk editing continues. We can't have similar response for the rest of actions just now, as this would break current UX, which relies on error response when at least one of rules fail. 

Intention is to build `update` action API  in a way it can handle this case(we also have it in new UX, displaying partial failures), then when building UI for it, adjust API for the rest of actions. So:

1) we will not break current UX
2) will adjust API responses to one format for every action

> But do we need to respond with 200 if some of the rule updates fail?

I think we do, if from hundreds of updated rules only one fails, sending error status would be misleading. In my opinion for bulk actions we should return 200, list of errors + count of processed rules. This way we would be able to notify user how many failures happened and how many rules were processed.

xcrzx replied:
Discussed this topic via Zoom, so posting a summary here for visibility:
Responding with 200 OK when we have failed rules will require adding logic on FE. We already have error handling logic, to simplify it looks like the following:

```ts
try {
  await performAcrion()
} catch (e) {
  showError(e)
}
```

That logic doesn't cover cases when we receive errors with 200 OK. So we have to update it something like this:

```ts
try {
  const response = await performAcrion()
  if (response.errors) {
    showError(errors)
  }
} catch (e) {
  showError(e)
}
```

In my opinion, an extra `response.error` check is redundant. Moreover, from an API user perspective, it's pretty easy to overlook that 200 OK could contain errors and not handle that case at all. So we can avoid those pitfalls by returning 500 if any rule updates fail.

vitaliidm replied:
changed it according to our discussion

vitaliidm commented on x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.test.ts:
```diff
@@ -56,13 +57,55 @@ describe('perform_bulk_action_schema', () => {
+    const message = retrieveValidationMessage(payload as PerformBulkActionSchema);
+
+    expect(getPaths(left(message.errors))).toEqual([
+      'Invalid value "update" supplied to "action"',
```
I think it happens because `updates` property is missing. So, it might be validation function treats `BulkAction.update` as invalid as it can be valid only if `updates` property is present

xcrzx replied:
Hmm, this error message looks a bit misleading to me. Does `io-ts` allow to adjust it somehow?

banderror replied:
++ for adjusting it if it's possible

vitaliidm replied:
Doesn't look that proposed by @xcrzx [library](https://github.com/klesun/io-ts-better-union-error-reporter) helps to eliminate additional error messages:

So, on payload
```Javascript
{
    query: 'name: test',
    action: BulkAction.edit,
}
```

Our current error messages:

```Javascript
[
  'Invalid value "edit" supplied to "action"',
  'Invalid value "undefined" supplied to "edit"',
]
```

Validation errors with newly proposed library:

```
must satisfy either of
      | invalid 1 ExactType element(s)
        action: must satisfy either of
          | 'delete' expected, but 'edit' found
          | 'disable' expected, but 'edit' found
          | 'duplicate' expected, but 'edit' found
          | 'enable' expected, but 'edit' found
          | 'export' expected, but 'edit' found
      | invalid 1 ExactType element(s)
        edit: object[] is mandatory
```

Still, there is present a message regarding invalid action.

Possible solutions?

1. What if we put `edit` property as partial? 

```
export const performBulkActionSchema = t.exact(
  t.type({
    query: queryOrUndefined,
    action: bulkAction,
    [BulkAction.edit]: t.union([t.array(bulkActionEditPayload), t.undefined]),
  })
);
```
We can catch if edit is supplied with other actions(in body of route validation function) and throw error. 

Downsides: 
 - validation logic in 2 places

2. Dig deeper into union validation/error reporting, write own path reporter(??).  In that case I would suggest to handle it in a separate issue.

What are your thoughts @banderror , @spong , @xcrzx?

banderror replied:
This doesn't look anywhere better 🙂 

```
must satisfy either of
      | invalid 1 ExactType element(s)
        action: must satisfy either of
          | 'delete' expected, but 'edit' found
          | 'disable' expected, but 'edit' found
          | 'duplicate' expected, but 'edit' found
          | 'enable' expected, but 'edit' found
          | 'export' expected, but 'edit' found
      | invalid 1 ExactType element(s)
        edit: object[] is mandatory
```

Imho validation error messages should be clear for any user who doesn't fully understand the request/response schema and of course TS union types or any programming concepts like that. The whole `invalid 1 ExactType element(s)` line is just noise for the user.

> In that case I would suggest to handle it in a separate issue.

I agree with that. @vitaliidm could you please write up and open a ticket?

vitaliidm replied:
@banderror https://github.com/elastic/kibana/issues/121860

banderror commented on x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.ts:
```diff
@@ -6,13 +6,28 @@
+    t.type({
+      query: queryOrUndefined,
+      action: t.literal(BulkAction.update),
+      updates: t.array(bulkActionUpdate),
```
Perhaps for simplicity it'd make sense to call the property for the action's payload the same way as the action's name itself?

```ts
action: t.literal(BulkAction.update),
[BulkAction.update]: t.array(bulkActionUpdate),
```

vitaliidm replied:
changed as well

banderror commented on x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.ts:
```diff
@@ -6,13 +6,28 @@
+export const performBulkActionSchema = t.union([
+  t.exact(
+    t.type({
+      query: queryOrUndefined,
```
`query` is a common parameter for all the bulk actions, should it be lifted up in the schema?

vitaliidm replied:
lifted up ^,  thanks for the suggestion

banderror commented on x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.ts:
```diff
@@ -6,13 +6,28 @@
+  ),
+  t.exact(
+    t.type({
+      query: queryOrUndefined,
```
We will need to support applying bulk edits by a list of rule IDs, just double-checking is the plan to add this in a follow-up PR?

vitaliidm replied:
in follow-up PRs, when will be building UI, so ids could be passed when user select rules on a page

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.test.ts:
```diff
@@ -94,10 +96,38 @@ describe.each([
+
+      expect(response.status).toEqual(500);
       expect(response.body).toEqual({
-        message: 'mocked validation message',
-        status_code: 403,
+        message:
+          "Failed actions: 1. Rules processed: 3. 'Rule that fails': 'mocked validation message'",
+        status_code: 500,
```
Can we add some data about succeeded and failed updates to the response body? I think you had it in some previous commit, and it looked like that:

```ts
      expect(response.body).toEqual({
        errors: [
          {
            error: {
              message: 'mocked validation message',
              statusCode: 403,
            },
            rule: {
              id: mockRule.id,
              name: mockRule.name,
            },
          },
        ],
        failed_rules_count: 1,
        rules_count: 1,
        success: false,
      });
```

I'd suggest something like that:

```ts
      expect(response.body).toEqual({
        status_code: 500,
        message: 'Applying bulk edits partially failed',
        rules: {
          requested: 3,
          processed: 3,
          succeeded: 2,
          failed: 1,
        },
        errors: [
          {
            rule_id: mockRule.id,
            rule_name: mockRule.name,
            error_message: 'mocked validation message',
            error_status_code: 403,
          },
        ],
      });
```

`errors` array should probably be explicitly truncated to a large value like 1000 or 10000, as well as strings like `error_message` and `rule_name`.

vitaliidm replied:
implemented

banderror replied:
> number of rules limited to 10000

That's a good point, it could be sufficient. I just wasn't sure that it makes sense to return all the errors especially if there's a large number of them. While 10k is still a large number, it could be okish in this case.

> Do we limit any properties sizes in case of find_rules, where response is expected to be much larger than in this bulk edit errors response?

I don't think so. But `rules/_find` is a completely different endpoint that must return the data as it is. Here we're discussing a potentially large array of errors containing data that could be helpful for debugging, and this data doesn't have to be 100% precise or complete imho.

> to be safe, limiting error_message might be a good idea

👍

vitaliidm replied:
I've added limiting size of error message.

banderror reviewed (commented):
Partially reviewed the code and left some comments. Will continue with review and testing in a few hours. Thank you @vitaliidm!

vitaliidm commented on x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts:
```diff
@@ -453,6 +453,53 @@ export enum BulkAction {
+  'set_tags' = 'set_tags',
+  'add_index' = 'add_index',
+  'delete_index' = 'delete_index',
+  'set_index' = 'set_index',
```
changed

vitaliidm commented on x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts:
```diff
@@ -453,6 +453,53 @@ export enum BulkAction {
+  type: t.literal(BulkActionUpdateType.set_timeline),
+  value: t.type({
+    timelineId: timeline_id,
+    timelineTitle: timeline_title,
```
thanks for pointing it out. Chaged it as well

vitaliidm commented on x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts:
```diff
@@ -453,6 +453,53 @@ export enum BulkAction {
   'export' = 'export',
   'delete' = 'delete',
   'duplicate' = 'duplicate',
+  'update' = 'update',
```
I actually have been thinking to name it `edit`, but we already operate with name `update` when modifying rules(update APIs), so decided to go with `update` to be consistent.
Similarly, there are already defined bulk actions like `delete` that matches existing APIs(_bulk_delete).

Would having `edit` be preferable?

banderror replied:
Yeah, I think it would be preferable. This type of modification is different from the existing bulk update API (PUT semantics) and bulk patch API (PATCH semantics) - it implements idempotent adding and removal and could implement even more complicated logic in the future if we need it (off the top of my head - something like reset to initial params for modified prebuilt rules). In other words, bulk edit semantics != bulk update semantics. The reason I'm asking to not use the `update` name for this action is because I think it makes sense to reserve it for the simple bulk update logic if we make it a part of the bulk actions endpoint in the future.

cc @xcrzx @spong

vitaliidm replied:
renamed as `edit`

spong replied:
A little late and this has already been _updated_ (😉) , but +1 to `edit` semantics here. We'll probably want to touch base with @XavierM and the RAM (Rules/Alerts Mgmt) folks here soon as this is definitely something we'll want to fold into the existing alerting platform API (and so any feedback they may have would be useful at this point).

banderror replied:
@spong ++, we'll chat with Xavier on the future of Rule Management (including the APIs) in early January after all the holidays. Xavier is on PTO until the next year.

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/rules/bulk_action_update.ts:
```diff
@@ -0,0 +1,61 @@
+  return arr.filter((item) => !itemsSet.has(item));
+};
+
+export const appplyBulkActionUpdateToRule = (
+  existingRule: Partial<RulesSchema>,
+  action: BulkActionUpdatePayload
+): Partial<RulesSchema> => {
```
`RulesSchema` is the schema used in the API responses, and I think we should use it only to populate the responses. I would keep business logic like this function dependent on internal types like Alerting Framework's types, SO types, SO parameters etc. This would allow us to modularize the codebase and keep API schemas separate from the core domain logic (which is currently not super explicit and lives in routes, executors and all sorts of utilities).

Specifically, I think this function could depend on `RuleAlertType` exported from `x-pack/plugins/security_solution/server/lib/detection_engine/rules/types.ts`.

vitaliidm replied:
renamed as `edit`

vitaliidm replied:
Implemented: I'm passing now `RuleAlertType`.

banderror commented on x-pack/plugins/security_solution/server/lib/detection_engine/rules/utils.ts:
```diff
@@ -98,7 +98,7 @@ export interface UpdateProperties {
   meta: MetaOrUndefined;
   machineLearningJobId: MachineLearningJobIdOrUndefined;
-  filters: PartialFilter[];
+  filters: PartialFilter[] | undefined;
```
Is this change intentional?

vitaliidm replied:
Yes, it's intentional. 
`UpdateProperties` is used for calculateVersion, which checks supplied properties in patchRules.
If any of parameter properties not undefined, it will increment version. 

As I made `PatchRulesOptions` optional, I need to also be able to pass filters as undefined. If it's not present, version will not be incremented

banderror commented on x-pack/plugins/security_solution/server/utils/promises_pool.ts:
```diff
@@ -0,0 +1,50 @@
+ * 2.0.
+ */
+
+// TODO: taken from https://github.com/elastic/kibana/pull/121110 PR, ensure its not duplicated after merge
```
That's very-very cool stuff, and I see why it is better than a simple chunk-and-promise-all approach, but could we please cover it with unit tests in either of the PRs? 🙏 
cc @xcrzx

xcrzx replied:
Covered in my PR

banderror reviewed (commented):
A few more comments. Sorry @vitaliidm!

xcrzx commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts:
```diff
@@ -14,16 +14,61 @@ import { performBulkActionSchema } from '../../../../../common/detection_engine/
+import { FindResult } from '../../../../../../alerting/server';
 
 const BULK_ACTION_RULES_LIMIT = 10000;
+const BULK_ACTION_CONCURRENCY = 500;
```
++ 500 parallel requests to ES is indeed a lot. I've recently observed a couple of Elasticsearch OOMs due to a large number of incoming requests. So we'd better play safe and lower the number to 50.

vitaliidm replied:
thanks for a useful link with explanation of chosen concurrency number.
I moved it to constants, so we will keep in one place

vitaliidm commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts:
```diff
@@ -145,9 +187,51 @@ export const performBulkActionRoute = (
+            processingResponse = await chunkifyRulesAction(rules.data, async (rule) => {
+              throwHttpError(await mlAuthz.validateRuleType(rule.params.type));
+
+              const updatedRule = body.update.reduce(
+                (acc, action) => appplyBulkActionUpdateToRule(acc, action),
+                transformAlertToRule(rule)
+              );
```
That's interesting point, thank you @banderror 

I've designed it's with an idea that array actions just aggregations of possible single requests.
So either you can do something like this

POST update/add_tags
POST update/remove_tags
POST update/set_tags

OR that would be equal to

POST [update/add_tags, update/remove_tags, update/set_tags]

In both situations, result would be the same. It's predictable as actions are applying in historical order, and can rewrite each other, which is fine in my opinion as order is explicit and result should be intuitive.

On the other hand, defining order of applied operations by priority can be not very explicit, users will need to have priority map in mind when constructing requests.

Issue here could be only when user starts to put the same value in `add_tags`, `remove_tags`, or use these 2 actions together with `set_tags`. Maybe worth cover this in documentation that order of actions are applying in historical order

banderror replied:
> In both situations, result would be the same. It's predictable as actions are applying in historical order, and can rewrite each other, which is fine in my opinion as order is explicit and result should be intuitive.
> On the other hand, defining order of applied operations by priority can be not very explicit, users will need to have priority map in mind when constructing requests.

That makes total sense, I agree with you @vitaliidm 👍 Forget my suggestion 🙂 

> Maybe worth cover this in documentation that order of actions are applying in historical order

Hmm, probably yes 🤔 If we document that individual edits are applied in the order of specifying them in the request body, it will bring more clarity for the users. On the other hand, we will have to stick to this behaviour because otherwise changing it later would represent a breaking change in the documented behaviour. Which option are you leaning towards?

vitaliidm replied:
I think it worth to document it as we give users an opportunity to apply multiple edit actions per request, which can save traffic, time needed to apply bulk edit and simplify code.

banderror replied:
Sounds good to me 👍 In that case, let's please add it to the description of the endpoint in https://github.com/elastic/security-docs/pull/1317

vitaliidm commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts:
```diff
@@ -14,16 +14,61 @@ import { performBulkActionSchema } from '../../../../../common/detection_engine/
+    concurrency: BULK_ACTION_CONCURRENCY,
+    items: rules,
+    executor: async (rule) => actionPerformWrapper(async () => action(rule), rule),
+  });
```
done

vitaliidm commented on x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.test.ts:
```diff
@@ -56,13 +57,55 @@ describe('perform_bulk_action_schema', () => {
+    expect(message.schema).toEqual({});
+  });
+
+  test('updates property is valid when action is update', () => {
```
updated, thanks for the proposal

vitaliidm commented on x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts:
```diff
@@ -453,6 +453,53 @@ export enum BulkAction {
+  }),
+});
+
+export const bulkActionUpdate = t.union([
```
updated

vitaliidm commented:
@elasticmachine merge upstream

vitaliidm commented:
@elasticmachine merge upstream

banderror reviewed (commented):
This PR looks great and is super close to being finalized, thank you @vitaliidm for all the adjustments!
I left a few small comments and going to start testing the endpoint.

vitaliidm commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts:
```diff
@@ -145,9 +189,67 @@ export const performBulkActionRoute = (
+            },
+          };
+
+          return response.custom({
+            headers: {
+              'content-type': 'application/json',
+            },
+            body: Buffer.from(JSON.stringify(responseBody)),
+            statusCode: 500,
+          });
```
@banderror, I didn't use it because it expects payload as

```typescript
      message: string | Error;
      attributes?: ResponseErrorAttributes;
```

which is differs from what we agreed (doesn't even allow to pass in body status_code, we send with other error responses everywhere in security solution). So I decided to stick with `response.custom`

banderror replied:
Discussed over zoom.
@vitaliidm will open a ticket for tweaking `SiemResponseFactory.error` method so it could be used for returning arbitrary fields in the response. Then we'll be able to refactor this `response.custom` call.

banderror commented:
Tested locally - it seems to work as expected, no issues found 👍 

Performance-wise, this is the rough request-response roundtrip times I got on my machine (both ES and Kibana running locally):

```json
{
  "action": "edit",
  "query": "",
  "edit": [
    {
      "type": "add_tags",
      "value": ["sample_tag"]
    },
    {
      "type": "delete_tags",
      "value": ["sample_tag_1"]
    },
    {
      "type": "delete_index_patterns",
      "value": ["test-index-*"]
    },
    {
      "type": "add_index_patterns",
      "value": ["my-index-*"]
    }
  ]
}
```

- 2 rules: ~1.8 seconds
- 627 rules: ~15 seconds
- 1254 rules: ~28 seconds
- 2508 rules: ~55 seconds

So this grows almost linearly. I believe that in environments with network latency between ES and Kibana these number would be X times higher. This is something to keep in mind when designing and implementing the bulk editing UI - I don't think that showing a success or failure toast in 1 or 2 minutes after submitting the bulk edit, while the user have likely left the Rule Management page, would be a good UX.

cc @xcrzx

vitaliidm commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts:
```diff
@@ -145,9 +189,67 @@ export const performBulkActionRoute = (
         }
 
-        return response.ok({ body: { success: true, rules_count: rules.data.length } });
+        return response.ok({
+          body: {
+            success: true,
+            rules_count: rulesCount,
+          },
+        });
```
No, they are not used anywhere.

I think error and success responses should not necessarily have the same format, as they would be treated differently when a result of the response gets processed. I.e. when we handle error, we already know it's not successful.
Plus, there could be cases when we would just have a general error response

```javascript
{ message: 'Not Found', status_code: 404 }
```
or, so on

vitaliidm commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts:
```diff
@@ -145,9 +189,67 @@ export const performBulkActionRoute = (
               },
               body: responseBody,
             });
+          case BulkAction.edit:
+            processingResponse = await executeBulkAction(rules.data, async (rule) => {
+              throwHttpError(await mlAuthz.validateRuleType(rule.params.type));
```
@banderror , looks like there is already implemented optimization for ML check
https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/machine_learning/authz.ts#L44
so, it should call ML services only once, on mlAuthz initialization and then return cached value when calling `mlAuthz.validateRuleType`. 

Are you still keen to optimize further? How do you see then this optimization?

From top of my head:

First approach
1. store result of ML rule validation for ML type of rule as a separate value
2. when processing action, throw error when type is ML

we still would need to check in every action whether rule is ML, we'll need another function for that to wrap it, if we don't want just put there if clause. Essentially will have N calls for that new function or if statement

Second approach
1. Loop over all found rules firstly, and throw error if it is ML and fails validation, gather caught errors
2. Proceed with bulk action in another loop

In this case we will add another loop through rules. Again will have additional N calls and a bit more complicated logic

What do you think? Anything we can improve here?

banderror replied:
Discussed over zoom.
We don't need to optimize the performance because this call is cached, but it would make sense to normalize the response in such a way so that duplicate errors are de-duplicated, something like that:

```ts
          const responseBody = {
            message: errorsCount === rulesCount ? 'Bulk edit failed' : 'Bulk edit partially failed',
            status_code: 500,
            attributes: {
              errors: errors.map(({ error, rule }) => ({
                message:
                  error.message.length > MAX_ERROR_MESSAGE_LENGTH
                    ? error.message.slice(0, MAX_ERROR_MESSAGE_LENGTH)
                    : error.message,
                status_code: error.statusCode,
                rules: [{ id, name }]
              })),
              rules: {
                total: rulesCount,
                failed: errorsCount,
                succeeded: rulesCount - errorsCount,
              },
            },
          };
```

vitaliidm commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts:
```diff
@@ -145,9 +189,67 @@ export const performBulkActionRoute = (
+
+        if (errorsCount > 0) {
+          const responseBody = {
+            message: 'Bulk edit partially failed',
```
adjusted an error message

vitaliidm commented:
@elasticmachine merge upstream

banderror reviewed (approved):
There's a few small leftovers to do, but I won't be blocking this PR anymore.
The implementation is good and works well, thank you for addressing so many comments!

We also discussed that it'd probably make sense to add a feature switch for bulk editing (before 8.1 FF) if we realize that we're not 100% confident in this API's schema, behaviour, performance or stability. This should become clearer along the implementation of bulk editing UI.

LGTM 👍

banderror commented:
We probably need to increase the timeout for the bulk actions route handler like it's been done for the prepackaged rules endpoint:
https://github.com/elastic/kibana/pull/94531/files#diff-116035c943669e24e35b50f024896bad2ef127adb079469318d07ef7f7baf476R53-R59

xcrzx commented:
Checked out and tested locally with a large number of detection rules (5000). Here are a few things that I've noticed:

1. When a cluster is not under heavy load, bulk update of 5k rules completes in about two minutes. Not a great performance, but still usable.
2. When a cluster is under some load (I simulated that by sending 20 bulk update requests in parallel), almost all requests fail with 502 Bad Gateway after 10 minutes. So it looks like Kibana has a default 10 minutes timeout.
3. Even after responding with 502, the bulk update route handler continues to execute. ~30 minutes have passed already, but I still see log messages popping up from the route handled, so it still runs. So outgoing Elasticsearch requests don't get canceled. It concerns me, as it might "invisibly" overload our user's clusters. No visible actions happen, but Kibana is unresponsive; I cannot even open it locally.
4. I see tons of `[WARN ][plugins.alerting] rulesClient.update('d8fec0f0-63d9-11ec-b043-87dcce37fa3a') conflict, exceeded retries` in logs. Probably race conditions, needs further investigation.
5. A lot of `[Dmitriis-MacBook-Pro.local] block until refresh ran out of slots and forced a refresh: [BulkShardRequest [[.kibana_8.1.0_001][0]] containing [index {[.kibana_8.1.0][alert:745f8be1-63d7-11ec-b043-87dcce37fa3a], source[n/a, actual length: [2.4kb], max length: 2kb]}] blocking until refresh]` messages in Elasticsearch logs.

banderror commented:
Brainstormed the issue described above with @vitaliidm @xcrzx @banderror. Ideas:

1. Find out if it's possible to check from the route handler that its execution was cancelled by the router/kibana core.
1. Implement our own timeout logic in the route handler. E.g. after X minutes stop executing bulk updates and return a response with the details of what rules have been updated, failed, skipped.
1. Rate limiting on the route level (per Kibana instance per route). It should reject requests if the limit has been exceeded.
1. Chat with Kibana Core team on the problem and our ideas.
1. UI should not allow the user to send a lot of concurrent bulk actions - meaning that it should be blocking for the user unit the bulk request finishes (blocking modal window, blocking flyout).
1. Do a perf audit (using APM etc) of the existing `RulesClient.update`. Try to identify slow methods/requests and opportunities for optimization.
1. Optimize performance of the existing `RulesClient.update`.
1. Implement efficient bulk operations in the `RulesClient`: bulk get, bulk update. (existing ticket https://github.com/elastic/kibana/issues/53144)
1. Process bulk actions using a queue based on Task Manager.

vitaliidm commented:
@elasticmachine merge upstream

vitaliidm commented:
@banderror , @xcrzx 
A few updates:
> 1. Find out if it's possible to check from the route handler that its execution was cancelled by the router/kibana core.

Kibana request has events.completed$ observable, that will get fired when request either aborted or completed. I've implemented handling of it in one of the commits: https://github.com/elastic/kibana/pull/120472/commits/ffb568c4547d8cf2769cf229256db68af4fdbd2d
So now, abortSignal is getting passed to **promisePool** util and it will abort processing executions further, once request is aborted/completed. 

> 2. Rate limiting on the route level (per Kibana instance per route). It should reject requests if the limit has been exceeded.
There is already similar method in endpoint directory of security_solution plugin (https://github.com/elastic/kibana/blob/master/x-pack/plugins/security_solution/server/endpoint/routes/limited_concurrency.ts#L45)

Looks like it's what we need, it will return 429, when number of requests exceeds limit. 
But currently it works in a way that applied limit is shared between all routes, that configured to have rate limit.
For example, if we have 2 routes configured with limit 10. If first one takes 10 requests, a second one won't be accessible, and will return 429.

So, this method needs a refactoring:
1. Rate limit can be configured per route, per tag(?)
2. Limit should apply to a single route, not be shared with all configured routes

I working on refactoring and additional checking/testing. Would it be preferable to include rate limit part in this PR or maybe in a separate?

vitaliidm commented:
@elasticmachine merge upstream

vitaliidm commented:
@xcrzx , @banderror 
- added rate limit for bulk edit API
- moved limitedConcurrency to common plugin area from endpoint
- removed unused code in endpoint directory

Checked as well APM transactions

<img width="1249" alt="Screenshot 2022-01-04 at 17 45 01" src="https://user-images.githubusercontent.com/92328789/148213673-2769ea06-847b-4aad-af7e-1e736bd56187.png">

In patchRules method, rulesClient.update does 4 requests under the hood: authorization check, getting rule and 2 requests for update. 
As spotted during heavy load, all of these 4 requests could start to take much longer time:

<img width="1230" alt="Screenshot 2022-01-05 at 12 19 45" src="https://user-images.githubusercontent.com/92328789/148216683-c9f41ef0-c7cf-4067-b71d-9c67104ae88c.png">

Further optimization could be:

- reduce further number of parallel requests, when updating rules. From 50 to 30, 25, etc.

Having https://github.com/elastic/kibana/issues/53144 implemented, looks as the most preferable option in future.

xcrzx commented:
> With limited rate as 5, for ~ 600 rules, timing for requests are following:
> http.codes.500: ............................................. 4

@vitaliidm Why the response code is 500 with the rate limit enabled? Shouldn't it be 429?

vitaliidm commented:
> > With limited rate as 5, for ~ 600 rules, timing for requests are following:
> > http.codes.500: ............................................. 4
> 
> @vitaliidm Why the response code is 500 with the rate limit enabled? Shouldn't it be 429?

@xcrzx , I sent only 5 requests, which within rate limit. 500 error was because of conflict for some rules updates

If I start run with 10 requests in parallel for example:

```
--------------------------------------
Metrics for period to: 14:33:40(+0000) (width: 0.022s)
--------------------------------------

vusers.created_by_name.bulk edit: ........................... 10
vusers.created.total: ....................................... 10
http.request_rate: .......................................... 10/sec
http.requests: .............................................. 10


--------------------------------------
Metrics for period to: 14:33:50(+0000) (width: 0.019s)
--------------------------------------
http.codes.429: ............................................. 5
http.responses: ............................................. 5
http.response_time:
  min: ...................................................... 9718
  max: ...................................................... 9735
  median: ................................................... 9801.2
  p95: ...................................................... 9801.2
  p99: ...................................................... 9801.2


--------------------------------------
Metrics for period to: 14:34:20(+0000) (width: 5.446s)
--------------------------------------

http.codes.500: ............................................. 5
http.responses: ............................................. 5
http.response_time:
  min: ...................................................... 33041
  max: ...................................................... 38486
  median: ................................................... 36691.5
  p95: ...................................................... 37432.7
  p99: ...................................................... 37432.7

```

5 will be completed with 429
5 with 500

also, it's captured in functional test https://github.com/elastic/kibana/pull/120472/files#diff-033a39080b5edd054aecd3ddfa4a06fc9d7e145e71bda681a8bd3647efb21956R303

xcrzx reviewed (approved):
Tested the `_bulk_action` endpoint against concurrent load locally; everything seems to work fine, and Kibana doesn't hang anymore. Request abortion also works as expected. Thank you, Vitalii 👍

I added a small comment about unsubscription, not sure how it should work tbh, so maybe require further investigation.

vitaliidm commented:
@elasticmachine merge upstream

vitaliidm commented on x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts:
```diff
@@ -38,12 +142,20 @@ export const performBulkActionRoute = (
+
+      // subscribing to completed$, because it handles both cases when request was completed and aborted.
+      // when route is finished by timeout, aborted$ is not getting fired
+      request.events.completed$.subscribe(() => abortController.abort());
```
I think we should be fine and don't need to unsubscribe.
Under the hood completed$

```typescript
    const completed$ = merge<void, void>(finish$, aborted$).pipe(shareReplay(1), first());
```
If I not mistaken, `first` operator will take care of unsubscribing

pzl reviewed (approved):
approving to satisfy onboarding team block here
